ingestion/src/kldata.rs

```diff
@@ -0,0 +1,279 @@
+    sensor_and_level: Option<(i32, i32)>,
+}
+
+fn parse_meta_field<T: FromStr>(field: &str, expected_key: &'static str) -> Result<T, Error>
+where
+    <T as FromStr>::Err: Debug,
+{
+    let (key, value) = field.split_once('=').ok_or_else(|| {
+        Error::Parse(format!(
+            "unexpected field in kldata header format: {}",
+            field
+        ))
+    })?;
+
+    // TODO: Replace assert with error?
+    assert_eq!(key, expected_key);
+
+    let res = value
+        .parse::<T>()
+        .map_err(|_| Error::Parse(format!("invalid number in kldata header for key {}", key)))?;
+
+    Ok(res)
+}
+
+fn parse_meta(meta: &str) -> Result<(i32, i32, usize), Error> {
+    let mut parts = meta.splitn(4, '/');
+
+    let next_err = || Error::Parse("kldata header terminated early".to_string());
+
+    if parts.next().ok_or_else(next_err)? != "kldata" {
+        return Err(Error::Parse(
+            "kldata indicator missing or out of order".to_string(),
+        ));
+    }
+
+    let station_id = parse_meta_field(parts.next().ok_or_else(next_err)?, "nationalnr")?;
+    let type_id = parse_meta_field(parts.next().ok_or_else(next_err)?, "type")?;
+    let message_id = parse_meta_field(parts.next().ok_or_else(next_err)?, "messageid")?;
+
+    Ok((station_id, type_id, message_id))
+}
+
+fn parse_columns(cols_raw: &str) -> Result<Vec<ObsinnId>, Error> {
+    // this regex is taken from kvkafka's kldata parser
+    // let col_regex = Regex::new(r"([^(),]+)(\([0-9]+,[0-9]+\))?").unwrap();
+    // it is modified below to capture sensor and level separately, while keeping
+    // the block collectively optional
+    //
+    // TODO: is it possible to reuse this regex even more?
+    let col_regex = Regex::new(r"([^(),]+)\(([0-9]+),([0-9]+)\)?").unwrap();
+
+    col_regex
+        .captures_iter(cols_raw)
+        .map(|caps| match caps.len() {
+            2 => Ok(ObsinnId {
+                param_code: caps.get(1).unwrap().as_str().to_owned(),
+                sensor_and_level: None,
+            }),
+            4 => Ok(ObsinnId {
+                param_code: caps.get(1).unwrap().as_str().to_owned(),
+                sensor_and_level: Some((
+                    caps.get(2).unwrap().as_str().parse().unwrap(),
+                    caps.get(3).unwrap().as_str().parse().unwrap(),
+                )),
+            }),
+            _ => Err(Error::Parse("malformed entry in kldata column".to_string())),
+        })
+        .collect::<Result<Vec<ObsinnId>, Error>>()
+}
+
+fn parse_obs(csv_body: Lines<'_>, columns: &[ObsinnId]) -> Result<Vec<ObsinnObs>, Error> {
+    let mut obs = Vec::new();
+
+    for row in csv_body {
+        let (timestamp, vals) = {
+            let mut vals = row.split(',');
+
+            let raw_timestamp = vals
+                .next()
+                .ok_or_else(|| Error::Parse("empty row in kldata csv".to_string()))?;
+
+            // TODO: timestamp parsing needs to handle milliseconds and truncated timestamps?
```

Comment:
Comment to clarify what it is actually handling now?

Reply:
It assumes the time portion is HHMMSS, which is what is done in kvkafka, and all the sample messages Søren sent. Søren said something annoyingly scary and nonspecific about this maybe including milliseconds or being truncated. I'm going to hash out the details of that when testing against real messages, rather than write a hypothetical handling here, hence the TODO

Reply:
I honestly would rather just find the obsinn messages that break this in meaningful ways, and base my implementation off those instead of blindly copying. When I write the test cases I'm expecting this to break